### PR TITLE
Proposed extensions: textDocument/selectionRange

### PR DESCRIPTION
**Pull request by matklad**
Implementation for https://github.com/Microsoft/language-server-protocol/issues/613

### CODE REVIEW

**Comment by matklad** [comment]
Status: I've got the code to compile, but I haven't actually tested it with my server. Not sure how all proposed bits, in the VS Code and in the lsp client library, should fit together.

File: client/src/selectionRange.proposed.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+	export namespace languages {
+		export function registerSelectionRangeProvider(selector: DocumentSelector, provider: SelectionRangeProvider): Disposable;
+	}
+
+	//#endregion
+
+	//#region Joh - read/write in chunks
+
+	export interface FileSystemProvider {
```
Comment (rcjsuen) [inline client/src/selectionRange.proposed.ts]: What is this interface for?

File: protocol/src/protocol.selectionRange.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#### Selection Range Request (:leftwards_arrow_with_hook:)
+
+The selection range request is sent from the client to the server to return suggested selection ranges at a given position.
+The first range must contain the given position.
+Position can coincide with the start or end of the first range.
+Subsequent ranges must contain the previous range.
```
Comment (rcjsuen) [inline protocol/src/protocol.selectionRange.md]: I think more documentation is required to clarify what this request is supposed to do. I personally find the term "selection range" a little vague but maybe that's just me.

File: client/package.json [hunk 3]
```diff
@@ -24,7 +25,7 @@
 	},
 	"dependencies": {
 		"semver": "^5.5.0",
-		"vscode-languageserver-protocol": "3.14.1"
+		"vscode-languageserver-protocol": "file://../protocol"
```
Comment (dbaeumer) [inline client/package.json]: This shouldn't be necessary. After cloning `vscode-languageserver-node` and running npm install simply run `npm run symlink` which does a developement setup linking.

File: client/src/selectionRange.proposed.ts (hunk 1 above)
Comment (matklad) [reply client/src/selectionRange.proposed.ts]: Sloppy copy paste :sweat_smile:, removed!

**Comment by matklad** [comment]
Status update: I've successfully (after fiddling with npm quite a bit) built my extension with `selectionRanges` feature enabled. Will try to implement server side soon.

**Comment by matklad** [comment]
The server impl is read, and server and client successfully work together: https://github.com/rust-analyzer/rust-analyzer/commit/fffa0695dd7772efa82dbac652223bbdc51c62d3!

The only quirk is that current insider's does not have the latest version of proposed API (it uses Range instead of SelectionRange), so I had to tweak client library a bit to work with old API.

**Comment by matklad** [comment]
Sure! I believe this is currently blocked on VS Code side ([some questions there](https://github.com/Microsoft/language-server-protocol/issues/613#issuecomment-449080525)). @jrieken could you ping me once VS Code side is "finalized" (I understand that this is provisional API, hence quotes).

**Comment by matklad** [comment]
I've rebased and switched to the new API, but I am having hard-time actually testing my client with this code (I had success the time before this, but I can't reproduce it). Is there anything more to using my fork of `vscode-langaugeclient` than changing `dependencies` in my plugin to

```
    "dependencies": {
        "vscode-languageclient": "file:///home/matklad/projects/vscode-languageserver-node/client/"
    },
```

npm gives me some weird errors about extraneous packages :(

whyyyyy `npm` is not `cargo`? :(

**Comment by dbaeumer** [comment]
@matklad I usually use links. (e.g. I go into the node_modules folder delete `vscode-languageclient` and add a symlink pointing to the branch on disk).

**Comment by matklad** [comment]
@dbaeumer that did the trick, thanks!

I've verified that all the pieces work together, [including my server](https://github.com/rust-analyzer/rust-analyzer/commit/cd6519a749ec6e27f154f7261340538e782a73dd).

**Comment by dbaeumer** [comment]
I will merge this in and ship a next version as soon as we released the next VS Code stable version which contains the necessary proposed API.

**Comment by jrieken** [comment]
fyi - after implementing some provider and actually using them we got the feeling of going 180 of the single position vs multi position. The VS Code API will very likely be changed to accept multiple positions and return an array of ranges

**Comment by dbaeumer** [comment]
@jrieken thanks for the heads up.

@matklad can you already forsee this. We could still map this in the converter for now. At least the LSP API will not change.

**Comment by dbaeumer** [comment]
I merged the PR in and made new next version:

protocol@3.15.0-next.1
client@5.3.0-next.1
server@5.3.0-next.1

@aeschli @octref

**Comment by matklad** [comment]
@jrieken I am a little confused about VS Code state here. Are you switching to multi-position in the end? The current proposed API is still single-position.

**Comment by jrieken** [comment]
Yeah, switching to multiple positions is the current plan. Not sure why @dbaeumer already merged this...

**Comment by rcjsuen** [comment]
@matklad @jrieken It's probably @dbaeumer feels that the LSP API itself [will not change](https://github.com/Microsoft/vscode-languageserver-node/pull/441#issuecomment-457531809) so he would like adopters to start trying out and using the LSP API.

Or do we now anticipate the LSP API changing also?

**Comment by dbaeumer** [comment]
This got released as a next version (*-next.1) to get feedback on the LSP level as well. Normal adopters will not see that version and we are free to change the LSP API to multiple positions which I suggest to do.

**Comment by dbaeumer** [comment]
In addition the feature is in proposed state and users need to opt in using it.

**Comment by matklad** [comment]
@dbaeumer FYI, the vscode API changed again:

* `kind` is dropped
* parent-child relationship are now established using parent pointers

https://github.com/Microsoft/vscode/commit/b61768007d5c1e91f96096565ba60db3d04b4b86

**Comment by dbaeumer** [comment]
@matklad I noticed that as well. Are you willing to work on a adoption PR. We can merge it in as soon as we have another stabled release with the new API.

**Comment by matklad** [comment]
@dbaeumer yeah, I'll most likely handle this withing a week

**Comment by matklad** [comment]
done: https://github.com/Microsoft/vscode-languageserver-node/pull/474#pullrequestreview-219371694

**Pull request merged**